yt

```diff
@@ -62,159 +73,269 @@ max_fps="30"
-    playlistids=$(getresults "$pgrep" "(playlist)")
-    # if there are playlists or videos, append them to list
-    [ -n "$playlistids" ] && ids="$playlistids\n"
-    [ -n "$videoids" ] && ids="$ids$videoids"
-    # url prefix for videos and playlists
-    videolink="https://youtu.be/"
-    playlink="https://youtube.com/playlist?list="
-    # prompt the results to user infinitely until they exit (escape)
-    while true; do
-        choice=$(echo -e "$ids" | cut -d'	' -f2 | $promptcmd) # dont show id
-        if [ -z "$choice" ]; then exit; fi	# if esc-ed then exit
-        id=$(echo -e "$ids" | grep -Fwm1 "$choice" | cut -d'	' -f1) # get id of choice
-        echo -e "$choice\t($id)"
-        case $id in
-            # 11 digit id = video
-            ???????????) mpv "$videolink$id" $mpv_options --ytdl-format="$ytdlformats";;
-            # 34 digit id = playlist
-            ??????????????????????????????????) mpv "$playlink$id" $mpv_options --ytdl-format="$ytdlformats";;
-            *) exit ;;
-        esac
-    done
+  # if in search show query result vids
+  response="$(curl -s "https://www.youtube.com/results?search_query=$query" |
+    sed 's|\\.||g')"
+  # if unable to fetch the youtube results page, inform and exit
+  if ! grep -q "script" <<<"$response"; then
+    echo "unable to fetch yt"
+    exit 1
+  fi
+  # regex expression to match video and playlist entries from yt result page
+  vgrep='"videoRenderer":{"videoId":"\K.{11}".+?"text":".+?[^\\](?=")'
+  pgrep='"playlistRenderer":{"playlistId":"\K.{34}?","title":{"simpleText":".+?[^\"](?=")'
+  # grep the id and title
+  # return them in format id (type) title
+  getresults() {
+    grep -oP "$1" <<<"$response" |
+      awk -F\" -v p="$2" '{ print $1 "\t" p " " $NF}'
+  }
+  # get the list of videos/playlists and their ids in videoids and playlistids
+  videoids=$(getresults "$vgrep")
+  playlistids=$(getresults "$pgrep" "(playlist)")
+  # if there are playlists or videos, append them to list
+  [ -n "$playlistids" ] && ids="$playlistids\n"
+  [ -n "$videoids" ] && ids="$ids$videoids"
+  # url prefix for videos and playlists
+  videolink="https://youtu.be/"
+  playlink="https://youtube.com/playlist?list="
+  # prompt the results to user infinitely until they exit (escape)
+  while true; do
+    choice=$(echo -e "$ids" | cut -d'	' -f2 | $promptcmd)       # dont show id
+    if [ -z "$choice" ]; then exit; fi                          # if esc-ed then exit
+    id=$(echo -e "$ids" | grep -Fwm1 "$choice" | cut -d'	' -f1) # get id of choice
+    echo -e "$choice\t($id)"
+    case $id in
+    # 11 digit id = video
+    ???????????) 
+      url="$videolink$id"
+      save_to_history "$url" "$choice"
+      mpv "$url" $mpv_options --ytdl-format="$ytdlformats" ;;
+    # 34 digit id = playlist
+    ??????????????????????????????????) 
+      url="$playlink$id"
+      save_to_history "$url" "$choice"
+      mpv "$url" $mpv_options --ytdl-format="$ytdlformats" ;;
```

Comment:
Seems like we call `save_to_history` every time before calling `mpv`, in that case this too can be moved to a separate function to reduce repetion.